UAVCAN: Read servo telemetry over CAN bus and publish to uorb

TedObrien opened the pull request:
<!--

Thank you for your contribution!

Get early feedback through
- Dronecode Discord: https://discord.gg/dronecode
- PX4 Discuss: http://discuss.px4.io/
- opening a draft pr and sharing the link

-->

### Solved Problem
While integrating [Currawong CBS20 ](https://www.currawongeng.com/can-servo/) servos into an airframe I wanted to monitor their performance to check for overheating, binding and general performance. PX4 has the ability to log ESC telemetry, so I added the ability to log servo telemetry over UAVCAN too. 

### Solution
The currawong servos publish the following [telemetry messages ](https://dronecan.github.io/Specification/7._List_of_standard_data_types/) to the can bus at different frequencies:
```
uavcan.equipment.actuator.Status
uavcan.equipment.power.CircuitStatus
uavcan.equipment.device.Temperature
```

[MKS](https://mks-servo.com/MKS-DroneCAN-Intro) seem to publish the same message set and [hitek](http://support.hitecrcd.net:7700/index.php/CanServo/Protocol/UAVCAN) appears to publish the `uavcan.equipment.actuator.Status` so I'm hoping this is a general solution - but haven't tested them myself.

I created 2 callbacks to read the `actuator.Status` and `device.Temperature`  messages, and publish them to two uorb messages `servo_status` and `servo_temperature`. The uavcan messages are published at different rates so thought it better not to combine them into one uorb message. I have tried to follow the same style as the `esc_status` message for consistency and used the [uavcan esc telemetry implementation](https://github.com/PX4/PX4-Autopilot/blob/main/src/drivers/uavcan/actuators/esc.cpp) as a guide on how to implement this.

I am new to UAVCAN so any feedback is very welcome :)

![Screenshot from 2025-03-21 13-42-19](https://github.com/user-attachments/assets/aabffdf2-2a32-4a2d-97f3-f00984b329fe)


### Changelog Entry

```
Feature: Add support for servo telemetry over UAVCAN
```

### Alternatives
We could also ...

### Test coverage
Tested with a Skynode and holybro v6x.

### Context
Related links, screenshot before/after, video

dakejahl commented on src/drivers/uavcan/actuators/servo.cpp:
```diff
@@ -60,3 +88,64 @@ UavcanServoController::update_outputs(bool stop_motors, uint16_t outputs[MAX_ACT
+
+		_servo_temperature.counter += 1;
+		_servo_temperature.timestamp = hrt_absolute_time();
+		_servo_temperature_pub.publish(_servo_temperature);
```
this callback will handle the `uavcan.equipment.device.Temperature` message from any node that publishes it, but here you're publishing it as a ServoTemperature.msg. Renaming would make it more clear:

ServoTemperature.msg --> DronecanDeviceTemperature.msg

dakejahl reviewed (commented):
I like the idea. I think you should name the uORB messages to match the Dronecan naming to make it more explicit.

ServoReport.msg --> DronecanActuatorStatus.msg
DeviceTemperature.msg --> DronecanDeviceTemperature
ServoTemperature.msg --> DronecanDeviceTemperatureArray

dagar commented:
> I like the idea. I think you should name the uORB messages to match the Dronecan naming to make it more explicit.
> 
> ServoReport.msg --> DronecanActuatorStatus.msg DeviceTemperature.msg --> DronecanDeviceTemperature ServoTemperature.msg --> DronecanDeviceTemperatureArray

Unless it's super specific to DroneCAN we should strive for generic feedback messages so that we could ultimately use them as part of control allocation or even just arming requirements.

dakejahl commented:
I suggested renaming to Dronecan* since the messages contain dronecan specific fields and the messages largely mirror the dronecan messages. We can always change things down the road

sfuhrer commented on msg/CMakeLists.txt:
```diff
@@ -202,6 +203,9 @@ set(msg_files
 	SensorAirflow.msg
+	ServoReport.msg
+	ServoStatus.msg
+	ServoTemperature.msg
```
This seems very odd to me - temperature is a core state of the Servo, I would expect it in the ServoStatus.

TedObrien replied:
Hi @sfuhrer, thanks for the review.

I left them separate as they are different messages in droneCAN and published at different rates (2Hz and 10Hz with the servo I am using). Would you prefer they were combined like in [airspeed.cpp](https://github.com/PX4/PX4-Autopilot/blob/main/src/drivers/uavcan/sensors/airspeed.cpp)?

I could have the `servo_status_sub_cb` publishing the `servo_status` message and the `servo_temperature_sub_cb`  updating the temperature values when they arrive to be published by the `servo_status_sub_cb`.

@dakejahl raised a point that currently any droneCAN with a  device_id < 8  publishing a `device.temperature` message will be picked up by the callback atm, this may need handling if the messages are to be combined?

sfuhrer replied:
I think it should be combined yes.

TedObrien commented:
> I'm also in favor of keeping the status messages generic. They should be usable for Dronecan but also for proprietary CAN protocols etc. This replaces #19822

Didn't realise there was already a pr for this, thanks for sharing, I'll have a look through it.

TedObrien commented:
Thanks for the feedback. I have modified the following:  

- Added temperature, current and voltage to the servo_report message and removed the servo_temperature message. 

- Added the servo_count and servo_function fields like in the previous pr https://github.com/PX4/PX4-Autopilot/pull/19822.

- Included individual counters for temperature and power as I thought it useful to see when the messages are being updated from the various uavcan messages.

Unsure if the counters are the best way dong this. Would it be preferable to include the timestamps instead? Or publish NaN values if no new message have been received from a specific callback for some specified time?

sfuhrer commented:
This is still of relevance.
